spec/abci++/abci++_methods.md

```diff
@@ -364,24 +452,20 @@ title: Methods
       size in bytes exceeds `RequestPrepareProposal.max_tx_bytes`.
       Therefore, if the size of `RequestPrepareProposal.txs` is greater than
       `RequestPrepareProposal.max_tx_bytes`, the Application MUST remove transactions to ensure
-      that the `RequestPrepareProposal.max_tx_bytes` limit is respected by those transaction
-      records returned in `ResponsePrepareProposal.tx_records` that are marked as `UNMODIFIED` or
-      `ADDED`.
+      that the `RequestPrepareProposal.max_tx_bytes` limit is respected by those transactions
+      returned in `ResponsePrepareProposal.txs` .
     * As a result of executing the prepared proposal, the Application may produce block events or transaction events.
-      The Application must keep those events until a block is decided and then pass them on to Tendermint via
-      `ResponseFinalizeBlock`.
+      The Application must keep those events until a block is decided. It will then forward the events to the `BeginBlock-DeliverTx-EndBlock` functions depending on where each event should be placed. The events are then returned to Tendermint when the transactions are indeed executed.
     * As a sanity check, Tendermint will check the returned parameters for validity if the Application modified them.
-      In particular, `ResponsePrepareProposal.tx_records` will be deemed invalid if
-        * There is a duplicate transaction in the list.
-        * A new transaction is marked as `UNMODIFIED` or `REMOVED`.
-        * An existing transaction is marked as `ADDED`.
-        * A transaction is marked as `UNKNOWN`.
+      In particular, `ResponsePrepareProposal.txs` will be deemed invalid if there are duplicate transactions in the list. <!-- ToDo  this is currenetly not implemented! -->
```

Comment:
A few thoughts:

Previously, it was not possible to include the same transaction twice in a block because the mempool would not admit a transaction twice and would therefore not ever include it twice in a block. This is somewhat incidental though because of the mempool's interaction with the gosisp layer. The mempool has to assume that a duplicately received transaction is a copy of a previously received transaction and should be ignored. Additionally, Tendermint does not perform a validity check on the proposed block for duplicate transactions.

This isn't exactly true when the application can submit Txs for `PrepareProposal`. Non-idempotent transactions may actually be meaningfully duplicated within a block, e.g. `X+1`. Disallowing this behavior disallows the application from including identical non-idempotent transactions. If the application is generating the transactions itself, it may wish to do this: Trivially, a simple counter may increment the same value twice in a block. 

The snag here is that the KV indexer relies on the transaction _hash_ to uniquely identify Txs. Clients may retrieve the Txs contents by hash. However, this really indicates that all transactions should be unique across the entire chain if a chain wishes to not lose data on duplicate transaction contents. Additionally, this behavior doesn't indicate what we should do for transactions _within_ a block. Chains _can_ shoot themselves in the foot by executing the same Tx across blocks, so why force the constraint? (Alternatively, we should drop support for indexing by hash and require height and hash for the query). 

All this to say, in doing this we'd be introducing a validity constraint that isn't really necessary for the new `PrepareProposal` call and doesn't actually target the indexing problem that is produced with duplicate transactions. I'd say we forgo this check, unless there's something I'm not seeing.

Reply:
OK, our latest internal discussions seem to imply we don't really know whether the Apps want Tendermint to do this check or not. So I would suggest to skip it for the time being, and adapt the spec accordingly.

Actually, I would make clear in this part of the spec that we are NOT enforcing the check, in case some App might be assuming we do.